test/token/ERC20WithPermit.test.js

```diff
@@ -0,0 +1,1102 @@
+      const permitTypehash = await token.PERMIT_TYPEHASH()
+      const nonce = await token.nonces(permittingHolder.address)
+
+      const approvalDigest = ethers.utils.keccak256(
+        ethers.utils.solidityPack(
+          ["bytes1", "bytes1", "bytes32", "bytes32"],
+          [
+            "0x19",
+            "0x01",
+            domainSeparator,
+            ethers.utils.keccak256(
+              ethers.utils.defaultAbiCoder.encode(
+                [
+                  "bytes32",
+                  "address",
+                  "address",
+                  "uint256",
+                  "uint256",
+                  "uint256",
+                ],
+                [
+                  permitTypehash,
+                  permittingHolder.address,
+                  spender,
+                  amount,
+                  nonce,
+                  deadline,
+                ]
+              )
+            ),
+          ]
+        )
+      )
+
+      return ethers.utils.splitSignature(
+        await signingKey.signDigest(approvalDigest)
+      )
+    }
+
+    context("when permission expired", () => {
+      it("should revert", async () => {
+        const deadline = yesterday
+        const signature = await getApproval(
+          permittingHolderBalance,
+          anotherAccount.address,
+          deadline
+        )
+
+        await expect(
+          token
+            .connect(anotherAccount)
+            .permit(
+              permittingHolder.address,
+              anotherAccount.address,
+              permittingHolderBalance,
+              deadline,
+              signature.v,
+              signature.r,
+              signature.s
+            )
+        ).to.be.revertedWith("Permission expired")
+      })
+    })
+
+    context("when permission has an invalid signature", () => {
+      it("should revert", async () => {
+        const deadline = tomorrow
+        const signature = await getApproval(
+          permittingHolderBalance,
+          anotherAccount.address,
+          deadline
+        )
+
+        await expect(
+          token.connect(anotherAccount).permit(
+            anotherAccount.address, // does not match the signature
+            anotherAccount.address,
+            permittingHolderBalance,
+            deadline,
+            signature.v,
+            signature.r,
+            signature.s
+          )
+        ).to.be.revertedWith("Invalid signature")
+      })
+    })
+
+    context("when the spender is not the zero address", () => {
+      context("when the sender has enough balance", () => {
+        const allowance = permittingHolderBalance
+        it("should emit an approval event", async () => {
+          const deadline = tomorrow
+          const signature = await getApproval(
+            allowance,
+            anotherAccount.address,
+            deadline
+          )
+
+          const tx = await token
+            .connect(anotherAccount)
```

Comment:
Can we test that anyone (neither owner nor spender) can call `permit` function with a valid signature?